programs/token/src/instructions/set_authority.rs

```diff
@@ -0,0 +1,82 @@
+use core::mem::MaybeUninit;
+
+use pinocchio::{
+    account_info::AccountInfo, instruction::{AccountMeta, Instruction, Signer}, program::invoke_signed, pubkey::Pubkey, ProgramResult
+};
+
+#[repr(u8)]
+#[derive(Clone, Copy)]
+pub enum AuthorityType {
+    MintTokens,
+    FreezeAccount,
+    AccountOwner,
+    CloseAccount,
+}
+
+
+/// Sets a new authority of a mint or account.
+///
+/// ### Accounts:
+///   0. `[WRITE]` The mint or account to change the authority of.
+///   1. `[SIGNER]` The current authority of the mint or account.
+pub struct SetAuthority<'a> {
+    /// Account (Mint or Token)
+    pub account: &'a AccountInfo,
+
+    /// Authority of the Account.
+    pub authority: &'a AccountInfo,
+
+    /// The type of authority to update.
+    pub authority_type: AuthorityType,
+
+    /// The new authority
+    pub new_authority: Option<Pubkey>,
```

Comment:
I think it would be better to use a reference to `Pubkey` to avoid one extra copy when setting the value.